### ISSUE TITLE
k8s node-e2e tests [120/121]

### ISSUE THREAD

**Issue by mrunalp** [kubernetes-incubator/cri-o#441]
```
make test-e2e-node PARALLELISM=1 RUNTIME=remote CONTAINER_RUNTIME_ENDPOINT=/var/run/ocid.sock IMAGE_SERVICE_ENDPOINT=/var/run/ocid.sock TEST_ARGS="--prepull-images=true" FOCUS="\[Conformance\]"

...
[Fail] [k8s.io] Probing container [It] should be restarted with a exec "cat /tmp/health" liveness probe [Conformance] 
/root/gosrc/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/container_probe.go:404

[Fail] [k8s.io] Networking [k8s.io] Granular Checks: Pods [It] should function for intra-pod communication: udp [Conformance] 
/root/gosrc/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/networking_utils.go:217

[Fail] [k8s.io] Networking [k8s.io] Granular Checks: Pods [It] should function for node-pod communication: udp [Conformance] 
/root/gosrc/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/networking_utils.go:270

[Fail] [k8s.io] Networking [k8s.io] Granular Checks: Pods [It] should function for node-pod communication: http [Conformance] 
/root/gosrc/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/networking_utils.go:270

[Fail] [k8s.io] Networking [k8s.io] Granular Checks: Pods [It] should function for intra-pod communication: http [Conformance] 
/root/gosrc/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/networking_utils.go:217

[Fail] [k8s.io] KubeletManagedEtcHosts [It] should test kubelet managed /etc/hosts file [Conformance] 
/root/gosrc/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/exec_util.go:107
```

**Comment by mrunalp** [comment]
With #440 we fixed one of these, so only 11 remaining after #436 is merged.

**Comment by mrunalp** [comment]
A few of these are because we don't separate stderr/stdout. We can get that fixed after merging #436

**Comment by cyphar** [comment]
I presume the `stderr/stdout` split is for `terminal=false` containers? I think after #436 we can make the splitting work with minimal changes to my trick for trailing lines.

**Comment by mrunalp** [comment]
@cyphar Yes, for terminal=false containers.

**Comment by sameo** [comment]
@mrunalp I will look at the networking issues. Is there a way to only run specific tests ? `--with-test-args=""` ?

**Comment by runcom** [comment]
@sameo it's:
```
make test-e2e-node ..... FOCUS="test name \[Conformance\]"
```

**Comment by runcom** [comment]
Fixed all termination path tests here https://github.com/kubernetes-incubator/cri-o/pull/443

**Comment by runcom** [comment]
6 more to go I guess :sunglasses:

**Comment by runcom** [comment]
```
[Fail] [k8s.io] Projected [It] should project all components that make up the projection API [Conformance] [Volume] [Projection] 
/root/gosrc/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2213
```
this test in the first comment is fixed according to my testing

**Comment by runcom** [comment]
updated the first comment with just the tests failing

**Comment by sameo** [comment]
@runcom @mrunalp @cyphar So the networking tests are failing with this error:
```
Apr 19 12:26:17.910: INFO: Failed to execute "echo 'hostName' | timeout -t 2 nc -w 1 -u 10.88.4.202 8081 | grep -v '^\\s*$'": unable to upgrade connection: URL is not set, stdout: "", stderr: ""
```

Those 2 UDP and HTTP tests set a test container within the pod and are asking `kubectl` to `exec` some ncat or curl command from the test container.
Since we do not support container_exec yet, it fails.

I'm going to give @jjlakis PR a try and see if it helps.

**Comment by sameo** [comment]
So after doing 2 things:

1. Apply PR #282 
2. Disable firewalld (I need to investigate which of the FC25 firewalld rule is blocking 8080 and 8081 on cni0...)

I get all 4 of the networking test failures to pass.

Looking at PR #282, I think there are 2 areas for improvement:

1. Implement `ocic exec`
2. Use conmon instead of calling into the runtime directly.

I will work on 1, and I think we can live without 2 for now and track that as a separate issue.

**Comment by jjlakis** [comment]
@sameo I was thinking about implementing `ocic exec` in two ways as I mentioned in https://github.com/kubernetes-incubator/cri-o/pull/282#issuecomment-269604462 , consider this :) .
If you want me to rebase I can do it, but if you want to take my commits and use it in another PR with more stuff - feel free to do it! Viva cri-o!

**Comment by sameo** [comment]
@jjlakis Could you please reset your @intelsdi-x/exec branch and pull from https://github.com/sameo/cri-o/tree/exec ?

I rebased and implemented `ocic ctr exec` on top of your patches.

**Comment by mrunalp** [comment]
@sameo @jjlakis Awesome! I think for firewalld we can open up the ports in setup and disable them in teardown. We can handle conmon after getting this in. We will still need to handle the tty resize though before we merge #282. I'll comment there. 

@cyphar we'll still need changes to conmon for attach so we can fix it together.

**Comment by mrunalp** [comment]
I added changes for tty resize handling on top of @sameo branch and have kubectl exec with sh working :)
```
[root@localhost kubernetes]# ./cluster/kubectl.sh logs httpd-2942495395-xk61h
AH00558: httpd: Could not reliably determine the server's fully qualified domain name, using 10.88.3.202. Set the 'ServerName' directive globally to suppress this message
AH00558: httpd: Could not reliably determine the server's fully qualified domain name, using 10.88.3.202. Set the 'ServerName' directive globally to suppress this message
[Thu Apr 20 17:47:20.093224 2017] [mpm_event:notice] [pid 1:tid 140686631569280] AH00489: Apache/2.4.25 (Unix) configured -- resuming normal operations
[Thu Apr 20 17:47:20.093344 2017] [core:notice] [pid 1:tid 140686631569280] AH00094: Command line: 'httpd -D FOREGROUND'
[root@localhost kubernetes]# ./cluster/kubectl.sh exec httpd-2942495395-xk61h -i -t date
Thu Apr 20 17:47:42 UTC 2017
[root@localhost kubernetes]# ./cluster/kubectl.sh exec httpd-2942495395-xk61h  date
Thu Apr 20 17:47:50 UTC 2017
[root@localhost kubernetes]# ./cluster/kubectl.sh exec httpd-2942495395-xk61h  -i -t sh
# 
# ps -ef
UID        PID  PPID  C STIME TTY          TIME CMD
root         1     0  0 17:47 ?        00:00:00 httpd -DFOREGROUND
daemon       7     1  0 17:47 ?        00:00:00 httpd -DFOREGROUND
daemon       8     1  0 17:47 ?        00:00:00 httpd -DFOREGROUND
daemon       9     1  0 17:47 ?        00:00:00 httpd -DFOREGROUND
root       103     0  0 17:47 pts/0    00:00:00 sh
root       109   103  0 17:47 pts/0    00:00:00 ps -ef
# 
#   
```
I will push that commit to a branch.

**Comment by mrunalp** [comment]
Here is my branch with the commit on top https://github.com/mrunalp/cri-o/commits/exec_tty @sameo @jjlakis could you guys cherry pick it?

**Comment by sameo** [comment]
@mrunalp I cherry picked it, thanks.

@jjlakis https://github.com/sameo/cri-o/commits/exec contains all needed patches now.

**Comment by mrunalp** [comment]
With #466 and #468 we are down to just one failing test :tada: 

Let's get those in :)

**Comment by runcom** [comment]
Those tests are now failing on my F25 box when running k8s e2e:
```
should project all components that make up the projection API [Conformance] [Volume] [Projection] 
should be consumable from pods in volume as non-root with defaultMode and fsGroup set [Conformance] [Volume]
```

@sameo @mrunalp could you give these tests a try?

**Comment by mrunalp** [comment]
@runcom The seconds in your list passes for me. The first is a known problem at this point. I am looking into it.

**Comment by mrunalp** [comment]
I found the issue with `should project all components that make up the projection API [Conformance] [Volume] [Projection] ` kubelet isn't reading our log file because it isn't terminated with a `\n` and that is actually a kubelet bug. Adding a simple echo hi to the end of the command fixes it :)

**Comment by mrunalp** [comment]
```
[root@localhost kubernetes]# git diff
diff --git a/test/e2e/common/projected.go b/test/e2e/common/projected.go
index 8853c67..fc6469b 100644
--- a/test/e2e/common/projected.go
+++ b/test/e2e/common/projected.go
@@ -1005,7 +1005,7 @@ var _ = framework.KubeDescribe("Projected", func() {
                        {
                                Name:    "projected-all-volume-test",
                                Image:   "gcr.io/google_containers/busybox:1.24",
-                               Command: []string{"sh", "-c", "cat /all/podname && cat /all/secret-data && cat /all/configmap-data"},
+                               Command: []string{"sh", "-c", "cat /all/podname && cat /all/secret-data && cat /all/configmap-data && echo hi"},
                                VolumeMounts: []v1.VolumeMount{
                                        {
                                                Name:      "podinfo",
```

**Comment by mrunalp** [comment]
Upstream issue https://github.com/kubernetes/kubernetes/issues/44976

**Comment by mrunalp** [comment]
All tests passing :tada: 
```

Ran 121 of 213 Specs in 2957.196 seconds
SUCCESS! -- 121 Passed | 0 Failed | 0 Pending | 92 Skipped PASS
```

**Comment by rhatdan** [comment]
This is awesome.
👍

**Comment by runcom** [comment]
closing :tada:

**Comment by runcom** [comment]
actually, let's leave this open for https://github.com/kubernetes-incubator/cri-o/issues/441#issuecomment-297554943

**Comment by rhatdan** [comment]
@mrunalp  I think we can close this now?

**Comment by runcom** [comment]
Yup!
